### PR TITLE
fix(napi):  prevent access to tsfn-raw after env's destroyed(#1514)

### PR DESCRIPTION
**Pull request by HotQ**
Because the life cycle of tsfn-raw is bound to node env, so we can add tsfn-raw to cleanup hook as well.

### CODE REVIEW

File: crates/napi/src/status.rs [hunk 1]
```diff
@@ -30,6 +30,8 @@ pub enum Status {
   DetachableArraybufferExpected,
   WouldDeadlock,
   NoExternalBuffersAllowed,
+  /// The napi_env passed into the call was not valid
+  EnvInvalid,
```
Comment (Brooooooklyn) [inline crates/napi/src/status.rs]: Don't need an additional error code here, these statuses are aligned to https://nodejs.org/api/n-api.html#napi_status

File: crates/napi/src/threadsafe_function.rs [hunk 2]
```diff
@@ -97,23 +98,83 @@ type_level_enum! {
 }
 
 struct ThreadsafeFunctionHandle {
-  raw: sys::napi_threadsafe_function,
+  raw: AtomicPtr<sys::napi_threadsafe_function__>,
   aborted: RwLock<bool>,
+  valid: AtomicBool,
```
Comment (Brooooooklyn) [inline crates/napi/src/threadsafe_function.rs]: Don't need `valid` status here, reuse `aborted` is fine.

File: crates/napi/src/threadsafe_function.rs [hunk 3]
```diff
@@ -97,23 +98,83 @@ type_level_enum! {
...
   referred: AtomicBool,
 }
 
 unsafe impl Send for ThreadsafeFunctionHandle {}
 unsafe impl Sync for ThreadsafeFunctionHandle {}
 
+impl ThreadsafeFunctionHandle {
+  // called when the env is being destroyed
+  // then we can't call napi_****_threadsafe_function any more
+  unsafe extern "C" fn clearup(self_ptr: *mut c_void) {
```
Comment (Brooooooklyn) [inline crates/napi/src/threadsafe_function.rs]: move it out of the `impl` block

File: crates/napi/src/threadsafe_function.rs (hunk 2 above)
Comment (HotQ) [reply crates/napi/src/threadsafe_function.rs]: got it

File: crates/napi/src/threadsafe_function.rs (hunk 3 above)
Comment (HotQ) [reply crates/napi/src/threadsafe_function.rs]: Is it because of mangling or something?

Comment (Brooooooklyn) [reply crates/napi/src/threadsafe_function.rs]: This is mainly for consistency with the rest of the project where `extern "C" fn` is used. Other places that don't put it in `impl`...

**Comment by HotQ** [comment]
forget about cleanup hook, we can only use `thread_finalize_cb `to achieve this.

File: crates/napi/src/threadsafe_function.rs [hunk 4]
```diff
@@ -307,20 +337,19 @@ impl<T: 'static, ES: ErrorStrategy::T> ThreadsafeFunction<T, ES> {
         async_resource_name,
         max_queue_size,
         1,
-        ptr::null_mut(),
+        Arc::into_raw(Pin::into_inner(handle.clone()))
+          .cast_mut()
```
Comment (Brooooooklyn) [inline crates/napi/src/threadsafe_function.rs]: This API was added in `1.65.0`: https://doc.rust-lang.org/std/primitive.pointer.html#method.cast_mut
It breaks the MSRV

**Comment by Brooooooklyn** [comment]
@HotQ have you tested this branch with Electron yet? Does this patch fix your problem?

**Comment by Brooooooklyn** [comment]
@HotQ let's deal with the worker thread case in another PR, thanks!

**Comment by Brooooooklyn** [comment]
@HotQ IMO, in this case: 

```rust
// got lock_guard if not aborted
if let Some(_guard) = tsfn.lock(){
    tsfn.call(Ok(()), NonBlocking)
}

pub fn lock(&self) -> Option<RwLockReadGuard<bool>> {
  let aborted_guard = self
    .handle
    .aborted
    .read()
    .expect("Threadsafe Function aborted lock failed");
  (!*aborted_guard).then_some(aborted_guard)
}
```

You even don't need to `guard` the `aborted` before calling. The `status` returned is enough to determine if it is successful or not.

**Comment by HotQ** [comment]
tsfn's call is currently unprotected, it means cleanup and finalize will happen between will  judging whether abortedand tsfn being called——And that will cause a crash(the tsfn has been deleted inside nodejs will be dereferenced).
so I need a guard to prevent the main thread's finalizing and the worker thread's access to tsfn from happening one after another.

**Comment by HotQ** [comment]
Or 
```rust
pub fn call(&self, value: Result<T>, mode: ThreadsafeFunctionCallMode) -> Status {
  let aborted_guard = self
      .handle
      .aborted
      .read()
      .expect("Threadsafe Function aborted lock failed");
  if !*aborted_guard {
    unsafe { sys::napi_call_threadsafe_function( ... ).into() }
  }else{
    Status::Cancelled
  }
}
`

**Pull request merged**